File: src/main/java/com/pancisin/bookster/components/SitemapService.java
package com.pancisin.bookster.components;

import java.net.MalformedURLException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.pancisin.bookster.models.enums.PageState;
import com.pancisin.bookster.models.enums.Visibility;
import com.pancisin.bookster.repository.ConferenceRepository;
import com.pancisin.bookster.repository.EventRepository;
import com.pancisin.bookster.repository.PageRepository;
import com.redfin.sitemapgenerator.WebSitemapGenerator;
import com.redfin.sitemapgenerator.WebSitemapUrl;

@Service
public class SitemapService {
	private static final String BASE_URL = "https://convene.sk/#";

	@Autowired
	private EventRepository eventRepository;

	@Autowired
	private PageRepository pageRepository;

	@Autowired
	private ConferenceRepository conferenceRepository;

	private final String[] staticRoutes = { "about", "pricing", "faq", "terms", "privacy-policy", "events", "explore",
			"conferences" };

	public String createSitemap() throws MalformedURLException {
		WebSitemapGenerator sitemap = new WebSitemapGenerator(BASE_URL);

		List<WebSitemapUrl> pages = pageRepository.findAll().stream()
				.filter(p -> p.getState().equals(PageState.PUBLISHED) || p.getState().equals(PageState.BLOCKED)).map(page -> {
					try {
						String identifier = page.getSlug() == null || page.getSlug().equals("") ? page.getId().toString()
								: page.getSlug();
						String url = String.join("/", BASE_URL, "page", identifier);

						return new WebSitemapUrl(url);
					} catch (MalformedURLException e1) {
						e1.printStackTrace();
						return null;
					}
				}).filter(x -> x != null).collect(Collectors.toList());

		sitemap.addUrls(pages);

		List<WebSitemapUrl> events = eventRepository.findAll().stream().filter(e -> e.getVisibility() == Visibility.PUBLIC)
				.map(event -> {
					try {
						String url = String.join("/", BASE_URL, "event", event.getId().toString());
						return new WebSitemapUrl(url);
					} catch (MalformedURLException e1) {
						return null;
					}
				}).filter(x -> x != null).collect(Collectors.toList());

		sitemap.addUrls(events);

		List<WebSitemapUrl> conferences = conferenceRepository.findAll().stream()
				.filter(c -> c.getState().equals(PageState.PUBLISHED) || c.getState().equals(PageState.BLOCKED))
				.map(conference -> {
					try {
						String url = String.join("/", BASE_URL, "conference", conference.getId().toString());
						return new WebSitemapUrl(url);
					} catch (MalformedURLException e1) {
						return null;
					}
				}).filter(x -> x != null).collect(Collectors.toList());

		sitemap.addUrls(conferences);
		
		List<WebSitemapUrl> staticUrls = Arrays.asList(staticRoutes).stream().map(r -> {
			try {
				String url = String.join("/", BASE_URL, r);
				return new WebSitemapUrl(url);
			} catch (MalformedURLException e1) {
				return null;
			}
		}).filter(x -> x != null).collect(Collectors.toList());
		
		sitemap.addUrls(staticUrls);

		return String.join("", sitemap.writeAsStrings());
	}
}

File: src/main/kotlin/com/pancisin/bookster/services/SitemapService.kt
package com.pancisin.bookster.services

import java.net.MalformedURLException

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

import com.pancisin.bookster.model.enums.PageState
import com.pancisin.bookster.model.enums.Visibility
import com.pancisin.bookster.repository.ConferenceRepository
import com.pancisin.bookster.repository.EventRepository
import com.pancisin.bookster.repository.PageRepository
import com.redfin.sitemapgenerator.WebSitemapGenerator
import com.redfin.sitemapgenerator.WebSitemapUrl
import java.util.*

@Service
class SitemapService {
  private final val BASE_URL: String = "https://convene.sk"

  @Autowired
  lateinit var eventRepository: EventRepository

  @Autowired
  lateinit var pageRepository: PageRepository

  @Autowired
  lateinit var conferenceRepository: ConferenceRepository

  private val staticRoutes = arrayOf("about", "pricing", "faq", "terms", "privacy-policy", "events", "explore", "conferences")

  fun createSitemap(): String? {
    val sitemap = WebSitemapGenerator(BASE_URL)

    val pages = pageRepository.findAll().filter { it.state === PageState.PUBLISHED || it.state === PageState.BLOCKED }.mapNotNull { page ->
      try {
        val identifier = if (page.slug?.equals("") ?: true) page.id.toString() else page.slug
        WebSitemapUrl("/${BASE_URL}/page/${identifier}")
      } catch (ex: MalformedURLException) {
        null
      }
    }
    sitemap.addUrls(pages)

    val events = eventRepository.findAll().filter {
      it.visibility === Visibility.PUBLIC && it.state === PageState.PUBLISHED && it.date?.compareTo(Calendar.getInstance()) ?: -1 >= 0
    }.mapNotNull {
      try {
        WebSitemapUrl("/${BASE_URL}/event/${it?.id.toString()}")
      } catch (ex: MalformedURLException) {
        null
      }
    }
    sitemap.addUrls(events)

    val conferences = conferenceRepository.findAll().filter { it.state === PageState.PUBLISHED || it.state === PageState.BLOCKED }.map {
      try {
        WebSitemapUrl("/${BASE_URL}/conference/${it.id}")
      } catch (ex: MalformedURLException) {
        null
      }
    }
    sitemap.addUrls(conferences)

    val staticUrls = staticRoutes.map {
      try {
        WebSitemapUrl("/${BASE_URL}/${it}")
      } catch (ex: MalformedURLException) {
        null
      }
    }
    sitemap.addUrls(staticUrls)

    return sitemap.writeAsStrings().joinToString("")
  }
}
